be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala

```diff
@@ -86,3 +94,35 @@ object Monitor {
   final case class TriggerHeartbeat() extends Event
   private final case class DoNothing() extends Event
 }
+
+// This class watch the list of server peers config file and upon changes
+// tells connectionMediator about it
+private class FileMonitor(mediatorRef: ActorRef) extends Runnable {
+
+  // getParent to exclude the filename from the path, i.e get the config directory path
+  private val directory: Path = Path.of(serverPeersListPath).getParent
+  private val watchService: WatchService = directory.getFileSystem.newWatchService()
+  directory.register(watchService, ENTRY_CREATE, ENTRY_MODIFY)
+
+  override def run(): Unit = {
+    try {
+      while (!Thread.currentThread().isInterrupted) {
+        // Blocks until an event happen
+        val watchKey = watchService.take()
+
+        // For any event, read the file and send it
+        for (event <- watchKey.pollEvents().asScala.toList) {
+          if (serverPeersListPath.endsWith(event.context().toString)) {
+            mediatorRef ! ConnectionMediator.ConnectTo(readServerPeers())
+          }
+          watchKey.reset()
```

Comment:
Do you need to do that? The events are already fetched and the watchKey will be deleted by the end of this while loop's scope anyway...

Reply:
As I understood it, this line is mandatory to put the key back from `signalled` into the `ready` state. Even if the variable holding the ref goes out of scope, the key it points to seems to be living as long as the watch service is not cancelled. I am very not comfortable with this... Here is the doc:

Reply:
Ok that makes more sense now!
However I think this should be moved out of the `for` loop then

Reply:
I hadn't seen that I think... yes indeed